### ISSUE TITLE
Foreign key exception at many to many add

### ISSUE THREAD

**Issue by JakobFerdinand** [dotnet/efcore#13205]
When adding a new `entity` to a M/N connection in a special order EF Core throws a `DbUpdateException` with message `The INSERT statement conflicted with the FOREIGN KEY constraint...`

```csharp
public static T CreateAddedEntity<T>(this DbContext context)
    where T : class
{
    var entity = System.Activator.CreateInstance<T>();
    context.Add(entity);
    return entity;
}
```

```csharp
using (var context = new SampleDbContext(useLogging: false))
{
    var teacher = context.Teachers.FirstOrDefault();

    // Step 1.
    var connections = context.Teacher_Students.ToList();
    // Step 2.
    var student = context.CreateAddedEntity<Student>();

    student.Name = "Alex";
    var ts = context.CreateAddedEntity<Teacher_Student>();
    ts.Teacher = teacher;
    // Step 3.
    ts.TeacherId = teacher.Id;

    ts.Student = student;
    // Step 4.
    ts.StudentId = student.Id;

    WriteLine();
    context.SaveChanges();
}
```

The order of the four steps are important.
For example if you first call step 2 `var student = context.CreateAddedEntity<Student>();` and then step 1 `var connections = context.Teacher_Students.ToList();` everthing works great.  
Of if you just comment out step 1, 3 or 4 everything works great.

```
Exception message:
 The INSERT statement conflicted with the FOREIGN KEY constraint "FK_Teacher_Students_Students_StudentId". The conflict occurred in database "ReproEfCoreManyToManyAddNewEntityProblemSample", table "dbo.Students", column 'Id'.
```

### Steps to reproduce
Here is a [Sample Project](https://github.com/rzlsoftware/EfCoreManyToManyAddNewEntityProblemSample) to demonstrate the problem.

### Further technical details
EF Core version: 2.1.2
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows10 18.03
IDE: Visual Studio 2017 15.8.2

**Comment by ajcvickers** [comment]
The root cause of this is setting an FK value on the join table (which is also part of the primary key) to a value which is temporary but different from the current value. The value is obtained from the Added principal. When a different value is explicitly set on a property that currently has a shadow value then it is set to be non-temporary. Not doing this in all cases would be a breaking change, and also not desirable. We could do some checking to see if the value is an existing temporary value. However, I believe if we implement #12378, then I don't think this case will repro anymore. So for now, I am proposing that we do nothing now, and instead revisit after #12378.

Workarounds are easy:
* Don't set the FK values explicitly--it's not needed in scenarios like this
* If the value was temporary, then manually mark it as temporary again after setting it.

**Comment by AndriySvyryd** [comment]
Also currently when we use the value generator from the principal entity type we mark the generated value as non-temporary if that's what the generator produces. However this doesn't make sense (even though it might not matter in most cases).

**Comment by ajcvickers** [comment]
Verified that this was fixed by the changes made for #12378.
